### PR TITLE
ENH: Implement DWI equality explicitly

### LINKED ISSUE #95: `DWI` objects equality is broken
```
File <attrs generated eq eddymotion.dmri.DWI>:4, in __eq__(self, other)
      2 if other.__class__ is not self.__class__:
      3     return NotImplemented
----> 4 return  (
      5     self.dataobj,
      6     self.affine,
      7     self.brainmask,
      8     self.bzero,
      9     self.gradients,
     10     self.em_affines,
     11     self.fieldmap,
     12     self._filepath,
     13 ) == (
     14     other.dataobj,
     15     other.affine,
     16     other.brainmask,
     17     other.bzero,
     18     other.gradients,
     19     other.em_affines,
     20     other.fieldmap,
     21     other._filepath,
     22 )

ValueError: The truth value of an array with more than one element is ambiguous. Use a.any() or a.all()
```

Is this automagically created by the attr library? It apparently doesn't have enough magic to deal with array attributes.

### PR DESCRIPTION
**Pull request by jhlegarreta**
- BUG: Do not pass the nibabel instance variable to `get_fdata`
- ENH: Implement DWI equality explicitly

Resolves #95.

### CODE REVIEW

**Review by oesteban — DISMISSED** [review]

**Comment by jhlegarreta** [comment]
`"dwi1.h5"` does not exist AFAIK (or at least I put the name just to be different from `"dwi.h5"`, and thus the PR was opened as a draft. Is there a DWI file that is not `"dwi.h5"` among the testing data that is pulled?

**Comment by jhlegarreta** [comment]
@oesteban After having shared some more data (thanks) this can be finished. Now do we stick to the `h5` file format for now for everything? I am not sure whether everything can be stored in there.

File: src/eddymotion/data/dmri.py [hunk 1]
```diff
@@ -79,6 +80,27 @@ def __len__(self):
         """Obtain the number of high-*b* orientations."""
         return self.dataobj.shape[-1]
 
+    def __eq__(self, other: Any):
+        return isinstance(other, DWI) and (
+            self.dataobj,
+            self.affine,
```
Comment (effigies) [inline src/eddymotion/data/dmri.py:86]: If these are ndarrays, you're going to get issues:

```python
In [1]: import numpy as np

In [2]: (np.eye(4),) == (np.eye(4),)
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
Cell In[2], line 1
----> 1 (np.eye(4),) == (np.eye(4),)

ValueError: The truth value of an array with more than one element is ambiguous. Use a.any() or a.all()
```

Possibly you want something like:

```python
# Check scalars
if (self._filepath, ...) != (other._filepath):
    return False

# Check array-likes
for attr in ('dataobj', 'affine', ...):
    if not np.allclose(getattr(self, attr), getattr(other, attr)):
        return False

return True
```

Comment (jhlegarreta) [reply src/eddymotion/data/dmri.py:86]: True, thanks Chris, I'll have another pass at this once I add another test dataset, or read the existing one, modify it, write it back to disk during the test, and use it as the righthand side value.

Comment (effigies) [reply src/eddymotion/data/dmri.py:86]: Might be cleanest to do a round-trip test: Load from raw data inputs, serialize to HDF5, load back the HDF5 and then check equality. Then you're definitely not relying on h5py `__eq__`s hiding bugs.

Comment (jhlegarreta) [reply src/eddymotion/data/dmri.py:86]: OK, will follow that path. Thanks for the insight Chris.

Comment (jhlegarreta) [reply src/eddymotion/data/dmri.py]: Done in 8ce53d6.

File: test/test_dmri.py [hunk 2]
```diff
@@ -65,3 +140,55 @@ def test_load(datadir, tmp_path):
...
+        b0_file=b0_fname,
+        brainmask_file=brainmask_fname,
+        fmap_file=fieldmap_fname,
+        b0_thres=b0_thres,
+    )
+
+    hdf5_filename = tmp_path / "test_dwi.h5"
+    lh_dwi_h5.to_filename(hdf5_filename)
+
+    rh_dwi_h5 = load(hdf5_filename)
```
Comment (effigies) [inline test/test_dmri.py:192]: Intuitively, I saw "lh" and "rh" and thought hemispheres.

```suggestion
    dwi_obj = load(
        dwi_fname,
        gradients_file=gradients_fname,
        b0_file=b0_fname,
        brainmask_file=brainmask_fname,
        fmap_file=fieldmap_fname,
        b0_thres=b0_thres,
    )

    hdf5_filename = tmp_path / "test_dwi.h5"
    dwi_obj.to_filename(hdf5_filename)

    round_trip_dwi_obj = load(hdf5_filename)
```

File: test/test_dmri.py [hunk 3]
```diff
@@ -65,3 +140,55 @@ def test_load(datadir, tmp_path):
...
+
+    assert lh_dwi_h5 == rh_dwi_h5
```
Comment (effigies) [inline test/test_dmri.py:194]:
```suggestion
    # Symmetric equality
    assert dwi_obj == round_trip_dwi_obj
    assert round_trip_dwi_obj == dwi_obj
```

File: src/eddymotion/data/dmri.py [hunk 4]
```diff
@@ -40,34 +40,46 @@ def _data_repr(value):
...
+    else:
+        raise NotImplementedError(f"Not implemented for {type(lh)} and {type(rh)}")
+
+
 @attr.s(slots=True)
 class DWI:
     """Data representation structure for dMRI data."""
 
-    dataobj = attr.ib(default=None, repr=_data_repr)
+    dataobj = attr.ib(default=None, repr=_data_repr, eq=attr.cmp_using(eq=np.allclose))
```
[the remark this answers is not in the document]
Comment (jhlegarreta) [reply src/eddymotion/data/dmri.py]: :+1:. Changed in a97f879.

File: src/eddymotion/data/dmri.py [hunk 5]
```diff
@@ -40,34 +40,46 @@ def _data_repr(value):
     return f"<{'x'.join(str(v) for v in value.shape)} ({value.dtype})>"
 
 
+def _cmp(lh, rh):
+    if lh is None and rh is None:
+        return True
+    elif lh is None and rh is not None:
```
[the remark this answers is not in the document]
Comment (jhlegarreta) [reply src/eddymotion/data/dmri.py]: :+1:. Changed in a97f879.

File: test/test_dmri.py (hunk 2 above)
Comment (jhlegarreta) [reply test/test_dmri.py]: :+1:. Changed in a97f879.

File: test/test_dmri.py (hunk 3 above)
Comment (jhlegarreta) [reply test/test_dmri.py]: :+1:. Changed in a97f879.

File: src/eddymotion/data/dmri.py [hunk 6]
```diff
@@ -40,34 +40,46 @@ def _data_repr(value):
...
 
+def _cmp(lh, rh):
+    if lh is None and rh is None:
+        return True
+    elif lh is None or rh is None:
+        return False
+    elif isinstance(lh, np.ndarray) and isinstance(rh, np.ndarray):
+        return np.allclose(lh, rh)
+    else:
+        raise NotImplementedError(f"Not implemented for {type(lh)} and {type(rh)}")
```
Comment (oesteban) [inline src/eddymotion/data/dmri.py:51]: May we relay on the original equality definitions of objects?

```suggestion
    if isinstance(lh, np.ndarray) and isinstance(rh, np.ndarray):
        return np.allclose(lh, rh)
    
    return lh == rh
```

**Review by oesteban — DISMISSED** [review]
Minor nit pick

**Review by oesteban — APPROVED** [review]

**Pull request merged**